### PR TITLE
Enable more passes in ProfilingGraphExecutor

### PR DESCRIPTION
**Pull request by Krovatkin**

### CODE REVIEW

File: torch/csrc/jit/profiling_graph_executor_impl.cpp [hunk 1]
```diff
@@ -11,29 +18,65 @@ bool& getProfilingMode() {
   return profiling_mode;
 }
 
-static std::shared_ptr<Graph> prepareGraph(const std::shared_ptr<Graph>& graph) {
+std::shared_ptr<Graph> ProfilingGraphExecutorImpl::prepareGraph(
+    const std::shared_ptr<Graph>& graph,
+    Stack& stack) {
   auto g = graph->copy();
+  ArgumentSpec spec =
```
Comment (zdevito) [inline torch/csrc/jit/profiling_graph_executor_impl.cpp]: With the profiling graph executor, we no longer need ArgumentSpec, so we should not carry it over here. I'd like to delete it, as its complexity is becoming a maintenance burden.

Comment (zdevito) [reply torch/csrc/jit/profiling_graph_executor_impl.cpp]: This is fine for WIP code though, since I know this relates to the PropagateRequiresGrad pass. However, we should edit propagates requires grad so that it works with ProfiledTensorType as well and can read the grad information from there.

File: torch/csrc/jit/profiling_graph_executor_impl.cpp [hunk 2]
```diff
@@ -11,29 +18,65 @@ bool& getProfilingMode() {
...
+        g, getAutodiffSubgraphInlining() ? autodiffSubgraphNodeThreshold : 1);
+    for (Node* dnode : diff_nodes) {
+      auto diff_graph = std::move(dnode->g(attr::Subgraph));
+      Gradient gradient = differentiate(diff_graph);
+      // do not optimize DifferentiableGraphs, since
+      // ideally they will be profiled and then optimized separetely
+      // when their corresponding DifferentiableGraphOp is called
+      packGradient(gradient, dnode);
+    }
+    InlineAutodiffSubgraphs(
```
Comment (zdevito) [inline torch/csrc/jit/profiling_graph_executor_impl.cpp]: We need to make sure that this pass, which undoes the creation of too-small gradients, does not accidentally lose profiling information for these things, because it is going to remove the DifferentiableGraphs where the optimization for these things would have taken place.

**Review by zdevito — APPROVED** [review]
Looks good for code that moves ProfilingGraphExecutor towards something that uses the same optimization pipeline. See inline comments for suggestions for future PRs.

File: torch/csrc/jit/profiling_graph_executor_impl.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/profiling_graph_executor_impl.cpp]: I'll double check

File: torch/csrc/jit/profiling_graph_executor_impl.cpp (hunk 1 above)
Comment (Krovatkin) [reply torch/csrc/jit/profiling_graph_executor_impl.cpp]: Yes, I had to use it since it fills in `requires_grad` for inputs. I'll try to detangle `requires_grad` propagation in a separate PR.

File: torch/csrc/jit/profiling_graph_executor_impl.cpp [hunk 3]
```diff
@@ -11,29 +18,64 @@ bool& getProfilingMode() {
...
+std::shared_ptr<Graph> ProfilingGraphExecutorImpl::prepareGraph(
+    const std::shared_ptr<Graph>& graph,
+    Stack& stack) {
   auto g = graph->copy();
+  ArgumentSpec spec =
+      arg_spec_creator_.create(autograd::GradMode::is_enabled(), stack);
+  arg_spec_creator_.specializeTypes(*g, spec);
   runRequiredPasses(g);
+  PropagateRequiresGrad(g);
+  if (needsGradient(g)) {
```
Comment (apaszke) [inline torch/csrc/jit/profiling_graph_executor_impl.cpp]: Isn't this piece of code copied from the original executor?

File: torch/csrc/jit/profiling_graph_executor_impl.cpp [hunk 4]
```diff
@@ -11,29 +18,65 @@ bool& getProfilingMode() {
...
+
+  if (!pr_) {
+    pr_ = ProfilingRecord::instrumentGraph(prepareGraph(graph, stack));
+    profiling_plan_ = ExecutionPlan(pr_->profiled_graph_);
+    // fall-through
+  }
+
   if (!pr_->ready()) {
-    return profiling_plan_;
+    std::cout << "running profiling graph\n";
```
Comment (jamesr66a) [inline torch/csrc/jit/profiling_graph_executor_impl.cpp]: Is this supposed to be here?
